runtime/src/accounts_db.rs

```diff
@@ -15546,4 +15546,107 @@ pub mod tests {
+            let mut purged_stored_account_slots = AccountSlots::default();
+            let mut purged_slot_pubkeys = HashSet::default();
+            purged_slot_pubkeys.insert((slot1, pk1));
+            let mut reclaims = SlotList::default();
+            db.accounts_index.upsert(
+                slot1,
+                slot1,
+                &pk1,
+                &AccountSharedData::default(),
+                &AccountSecondaryIndexes::default(),
+                AccountInfo::default(),
+                &mut reclaims,
+                UpsertReclaim::IgnoreReclaims,
+            );
+            db.unref_accounts(purged_slot_pubkeys, &mut purged_stored_account_slots);
+            assert_eq!(
+                vec![(pk1, vec![slot1].into_iter().collect::<HashSet<_>>())],
+                purged_stored_account_slots.into_iter().collect::<Vec<_>>()
+            );
+            assert_eq!(db.accounts_index.ref_count_from_storage(&pk1), 0);
+        }
+        {
+            let db = AccountsDb::new_single_for_tests();
+            let mut purged_stored_account_slots = AccountSlots::default();
+            let mut purged_slot_pubkeys = HashSet::default();
+            let mut reclaims = SlotList::default();
+            // pk1 and pk2 both in slot1 and slot2, so each has refcount of 2
+            for slot in [slot1, slot2] {
+                for pk in [pk1, pk2] {
+                    db.accounts_index.upsert(
+                        slot,
+                        slot,
+                        &pk,
+                        &AccountSharedData::default(),
+                        &AccountSecondaryIndexes::default(),
+                        AccountInfo::default(),
+                        &mut reclaims,
+                        UpsertReclaim::IgnoreReclaims,
+                    );
+                }
+            }
+            // purge pk1 from both 1 and 2 and pk2 from slot 1
+            let purges = vec![(slot1, pk1), (slot1, pk2), (slot2, pk1)];
+            purges.into_iter().for_each(|(slot, pk)| {
+                purged_slot_pubkeys.insert((slot, pk));
+            });
+            db.unref_accounts(purged_slot_pubkeys, &mut purged_stored_account_slots);
+            for (pk, slots) in vec![(pk1, vec![slot1, slot2]), (pk2, vec![slot1])] {
+                let result = purged_stored_account_slots.remove(&pk).unwrap();
+                assert_eq!(result, slots.into_iter().collect::<HashSet<_>>());
+            }
+            assert!(purged_stored_account_slots.is_empty());
+            assert_eq!(db.accounts_index.ref_count_from_storage(&pk1), 0);
+            assert_eq!(db.accounts_index.ref_count_from_storage(&pk2), 1);
+        }
+    }
+
+    #[test]
+    fn test_many_unrefs() {
+        let db = AccountsDb::new_single_for_tests();
+        let mut purged_stored_account_slots = AccountSlots::default();
+        let mut reclaims = SlotList::default();
+        let pk1 = Pubkey::new(&[1; 32]);
+        let n = 10_001;
```

Comment:
don't think it's worth exposing the batch size outside the function, but maybe make a comment here about why we're doing 10001:

```suggestion
        let n = 10_001; // enough slots to have multiple batches
```